### ISSUE TITLE
Grounding numbers

### ISSUE THREAD

**Issue by psubotic** [souffle-lang/souffle#1294]
If I have a rule like: 

P(i) :- G(i-1). 

Souffle will report i is ungrounded. This can be annoying for developers.
I feel with some rewriting this could be handled.

**Comment by b-scholz** [comment]
I see - you would like to have the following transformation:
```
  P(i) :- G(i-1). 
=> 
  P(i) :- G(j), j = i-1. 
=> 
  P(i) :- G(j), i=j+1. 
=> 
  P(j+1) :- G(j). 
```

The problem is that a general solution might be hard. We would need to find a solution for a system of simultaneous equations over grounded and ungrounded variables. 
You would like to express ungrounded variables as a function over grounded variables. 

The art here is to understand which class of systems you would like to tackle. For example, you have a simple equation with a single grounded / ungrounded variable and a linear transformation etc.

**Comment by psubotic** [comment]
yeah i think for now a non-generalised would be ok - assuming only 1
variable in the equation.

I did the transformation by hand but when people use P its unnatural to
know what offset they need to consider

**Comment by b-scholz** [comment]
I guess the main problem is that people are trained in Prolog and not in Datalog, i.e., the direction of the argumentation changes.

**Comment by psubotic** [comment]
yeah esp if you see the ruleset as a library.

e.g., GetParameter(v, i) :- ....

might result in some trickery but the user just wants parameter 1 2 or 3.
Naturally thinking top down....

**Comment by b-scholz** [comment]
We are working on Version 2 and still have a lot to do for Version 2.0. 
After completion of Version 2, this would be a very nice feature for beginners.

**Comment by psubotic** [comment]
that would be great

**Comment by psubotic** [comment]
the other alternative is to leave souffle datalog for "experts" and provide
a dsl that compiles down to it and performs such transforms
